packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -656,6 +656,33 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   @protected
   void addToScene(ui.SceneBuilder builder);
 
+  // TODO(fzyzcjy): would be great to have "addToSceneWrapped & addToScene"
+  // renamed to "addToScene & performAddToScene", just like layout&performLayout
+  /// Add the layer to scene
+  void addToSceneWrapped(ui.SceneBuilder builder) {
+    ui.EngineLayer? previousEngineLayer;
+    assert(() {
+      previousEngineLayer = engineLayer;
+      return true;
+    }());
+
+    addToScene(builder);
+
+    assert(() {
+      assert(
+        previousEngineLayer == null || previousEngineLayer != engineLayer,
```

Comment:
I believe, `addToScene` is not required to change the value of `engineLayer`. It could in theory call `builder.addRetained` with the existing layer. That would fail now.

Reply:
Hmm it looks like `builder.addRetained` is called in `_addToSceneWithRetainedRendering`. And `_addToSceneWithRetainedRendering` calls `addToSceneWrapped`, so the check will not be triggered. 

Indeed, looking at `_addToSceneWithRetainedRendering`, it seems that `addToSceneWrapped` will only be called if `_needsAddToScene`. So, if a layer says it needs to add to scene, but indeed it does not, this looks like a logical problem, so should not happen (and should make assert fail).

By the way, CI passes so it seems current flutter codebase does not have this problem.